apps/src/dance/ai/AiBlockPreview.tsx

```diff
@@ -19,7 +19,23 @@ const AiBlockPreview: React.FunctionComponent<AiBlockPreviewProps> = ({
 }) => {
   const blockPreviewContainerRef = useRef<HTMLSpanElement>(null);
   const refTimer = useRef<number | null>(null);
+  const workspaceRef = useRef<Workspace | null>(null);
```

Comment:
Help me understand the use of a ref here (vs other hooks) -- I always think of refs as primarily used to reference some DOM node. React docs describe it as storing values "not needed for rendering" -- that's not totally true here, right (ie, we need it to render)?

Reply:
I've started using `useRef` more, and currently understand that it gives us a variable that lives between re-renders, and that doesn't itself cause a re-render when changed.  Given that updating a variable created with `useState` triggers an immediate re-render (even before subsequent updates happen in the same function!), we want to be judicious in our use of `useState`, limiting it to state changes that we know will trigger an immediate re-render.  `useRef` then gives us a way to store additional variables, which can indeed be consumed by the rendering process, without actually triggering a re-render.

Reply:
^ yep exactly! In transitioning from class components to functional components with hooks, refs became more of a replacement for instance variables, rather than just DOM node references. In this case, I think the distinction is that the workspace object 1) isn't strictly a value that drives rendering in the React tree directly (as opposed to some simple state value like `shouldShowButton`) and 2) does render something on the page, but does so outside of the React virtual DOM (instead via direct DOM manipulation).

Reply:
Thanks for explaining!